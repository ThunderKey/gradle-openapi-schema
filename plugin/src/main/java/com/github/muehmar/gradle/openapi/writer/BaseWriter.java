package com.github.muehmar.gradle.openapi.writer;

import java.util.stream.IntStream;

/**
 * Base implementation of a {@link Writer}. Extending classes should implement the {@link
 * Writer#close(String)} method
 */
abstract class BaseWriter implements Writer {
  private static final int DEFAULT_SPACES_PER_TAB = 2;
  private static final String NEWLINE_STRING = "\n";

  private final String tab;

  /** Contains the complete content written with this Writer. */
  protected final StringBuilder sb;

  private int tabs;
  private boolean newline;

  protected BaseWriter() {
    this(DEFAULT_SPACES_PER_TAB);
  }

  protected BaseWriter(int spacesPerIndent) {
    this.tab = new String(new char[spacesPerIndent]).replace("\0", " ");
    this.sb = new StringBuilder();

    this.tabs = 0;
    this.newline = true;
  }

  @Override
  public Writer print(String string, Object... args) {
    if (newline) {
      printTabs();
      newline = false;
    }
    sb.append(String.format(string, args));
    return this;
  }

  private void printTabs() {
    IntStream.range(0, tabs).forEach(i -> sb.append(tab));
  }

  @Override
  public Writer println() {
    sb.append(NEWLINE_STRING);
    tabs = 0;
    newline = true;
    return this;
  }

  @Override
  public Writer tab(int tabs) {
    this.tabs = tabs;
    return this;
  }

  @Override
  public Writer ref(String ref) {
    return this;
  }
}
