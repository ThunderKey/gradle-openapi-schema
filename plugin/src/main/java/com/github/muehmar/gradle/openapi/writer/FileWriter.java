package com.github.muehmar.gradle.openapi.writer;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

/** Writes the content to a file. */
public class FileWriter extends BaseWriter {

  private final String basePath;

  public FileWriter(String basePath) {
    this.basePath = basePath;
  }

  @Override
  public boolean close(String subPath) {
    try {
      final Path file = Paths.get(basePath).resolve(subPath);
      file.toFile().mkdirs();

      Files.delete(file);

      try (final PrintWriter writer =
          new PrintWriter(
              new OutputStreamWriter(
                  new FileOutputStream(file.toFile()), StandardCharsets.UTF_8))) {

        writer.append(sb.toString());
        writer.flush();
      }
      return false;
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }
  }
}
