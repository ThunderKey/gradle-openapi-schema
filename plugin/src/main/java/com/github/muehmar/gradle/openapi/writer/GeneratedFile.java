package com.github.muehmar.gradle.openapi.writer;

import java.nio.file.Path;
import lombok.Value;

@Value
public class GeneratedFile {
  Path file;
  String content;
}
