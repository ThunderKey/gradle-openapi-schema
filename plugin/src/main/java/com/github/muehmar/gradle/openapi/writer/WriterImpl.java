package com.github.muehmar.gradle.openapi.writer;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.stream.IntStream;

public class WriterImpl implements Writer {
  private static final int DEFAULT_SPACES_PER_TAB = 2;
  private static final String NEWLINE_STRING = "\n";

  private final String tab;
  private final String basePath;
  private final StringBuilder sb;

  private int tabs;
  private boolean newline;

  public WriterImpl(String basePath) {
    this(basePath, DEFAULT_SPACES_PER_TAB);
  }

  public WriterImpl(String basePath, int spacesPerIndent) {
    this.basePath = basePath;
    this.tab = new String(new char[spacesPerIndent]).replace("\0", " ");
    this.sb = new StringBuilder();

    this.tabs = 0;
    this.newline = true;
  }

  @Override
  public Writer print(String string, Object... args) {
    if (newline) {
      printTabs();
      newline = false;
    }
    sb.append(String.format(string, args));
    return this;
  }

  private void printTabs() {
    IntStream.range(0, tabs).forEach(i -> sb.append(tab));
  }

  @Override
  public Writer println() {
    sb.append(NEWLINE_STRING);
    tabs = 0;
    newline = true;
    return this;
  }

  @Override
  public Writer tab(int tabs) {
    this.tabs = tabs;
    return this;
  }

  @Override
  public Writer ref(String ref) {
    return this;
  }

  @Override
  public boolean close(String subPath) {
    try {
      final Path file = Paths.get(basePath).resolve(subPath);
      file.toFile().mkdirs();

      Files.delete(file);

      try (final PrintWriter writer =
          new PrintWriter(
              new OutputStreamWriter(
                  new FileOutputStream(file.toFile()), StandardCharsets.UTF_8))) {

        writer.append(sb.toString());
        writer.flush();
      }
      return false;
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }
  }
}
