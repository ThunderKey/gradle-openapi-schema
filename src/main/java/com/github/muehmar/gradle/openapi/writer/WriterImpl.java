package com.github.muehmar.gradle.openapi.writer;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.stream.IntStream;

public class WriterImpl implements Writer {
  private static final int DEFAULT_SPACES_PER_TAB = 2;
  private static final String NEWLINE_STRING = "\n";

  private final String tab;
  private final String path;
  private final StringBuilder sb;

  private int tabs;
  private boolean newline;

  public WriterImpl(String path) {
    this(path, DEFAULT_SPACES_PER_TAB);
  }

  public WriterImpl(String path, int spacesPerIndent) {
    this.path = path;
    this.tab = new String(new char[spacesPerIndent]).replace("\0", " ");
    this.sb = new StringBuilder();

    this.tabs = 0;
    this.newline = true;
  }

  @Override
  public Writer print(String string, Object... args) {
    if (newline) {
      printTabs();
      newline = false;
    }
    sb.append(String.format(string, args));
    return this;
  }

  private void printTabs() {
    IntStream.range(0, tabs).forEach(i -> sb.append(tab));
  }

  @Override
  public Writer println() {
    sb.append(NEWLINE_STRING);
    tabs = 0;
    newline = true;
    return this;
  }

  @Override
  public Writer tab(int tabs) {
    this.tabs = tabs;
    return this;
  }

  @Override
  public Writer ref(String ref) {
    return this;
  }

  @Override
  public boolean close() {
    try {
      final Path file = Paths.get(path);
      file.toFile().mkdirs();

      file.toFile().delete();

      try (final PrintWriter writer =
          new PrintWriter(
              new OutputStreamWriter(
                  new FileOutputStream(file.toFile()), StandardCharsets.UTF_8))) {

        writer.append(sb.toString());
        writer.flush();
      }
      return false;
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }
  }
}
