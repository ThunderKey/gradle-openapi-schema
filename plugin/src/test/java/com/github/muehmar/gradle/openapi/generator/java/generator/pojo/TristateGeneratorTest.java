package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.generator.data.VoidData.noData;
import static com.github.muehmar.gradle.openapi.generator.java.generator.data.VoidData.noSettings;
import static org.junit.jupiter.api.Assertions.assertEquals;

import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;

class TristateGeneratorTest {
  @Test
  void tristateClass_when_called_then_correctOutput() {
    final Generator<Void, Void> gen = TristateGenerator.tristateClass();
    final Writer writer = gen.generate(noData(), noSettings(), Writer.createDefault());
    assertEquals(
        "package com.github.muehmar.openapi.util;\n"
            + "\n"
            + "import java.util.Objects;\n"
            + "import java.util.Optional;\n"
            + "import java.util.function.Function;\n"
            + "import java.util.function.Supplier;\n"
            + "\n"
            + "public final class Tristate<T> {\n"
            + "  private final Optional<T> value;\n"
            + "  private final boolean isNull;\n"
            + "\n"
            + "  private Tristate(Optional<T> value, boolean isNull) {\n"
            + "    this.value = value;\n"
            + "    this.isNull = isNull;\n"
            + "  }\n"
            + "\n"
            + "  /**\n"
            + "   * Returns a Tristate class whose value is mapped with the given function.\n"
            + "   */\n"
            + "  public <R> Tristate<R> map(Function<T, R> f) {\n"
            + "    return new Tristate<>(value.map(f), isNull);\n"
            + "  }\n"
            + "\n"
            + "  public static <T> Tristate<T> ofNullableAndNullFlag(T nullableValue, boolean isNull) {\n"
            + "    return new Tristate<>(Optional.ofNullable(nullableValue), isNull);\n"
            + "  }\n"
            + "\n"
            + "  public static <T> Tristate<T> ofValue(T value) {\n"
            + "    return new Tristate<>(Optional.of(value), false);\n"
            + "  }\n"
            + "\n"
            + "  public static <T> Tristate<T> ofAbsent() {\n"
            + "    return new Tristate<>(Optional.empty(), false);\n"
            + "  }\n"
            + "\n"
            + "  public static <T> Tristate<T> ofNull() {\n"
            + "    return new Tristate<>(Optional.empty(), true);\n"
            + "  }\n"
            + "\n"
            + "  /**\n"
            + "   * Registers a {@link java.util.function.Function} which is applied on the value\n"
            + "   * of the property if it was present and non-null.\n"
            + "   */\n"
            + "  public <R> OnNull<R> onValue(Function<T, R> onValue) {\n"
            + "    return onNull ->\n"
            + "      onAbsent -> value.map(onValue).orElseGet(() -> isNull ? onNull.get() : onAbsent.get());\n"
            + "  }\n"
            + "\n"
            + "  @Override\n"
            + "  public boolean equals(Object o) {\n"
            + "    if (this == o) return true;\n"
            + "    if (o == null || getClass() != o.getClass()) return false;\n"
            + "    Tristate<?> tristate = (Tristate<?>) o;\n"
            + "    return isNull == tristate.isNull && Objects.equals(value, tristate.value);\n"
            + "  }\n"
            + "\n"
            + "  @Override\n"
            + "  public int hashCode() {\n"
            + "    return Objects.hash(value, isNull);\n"
            + "  }\n"
            + "\n"
            + "  @Override\n"
            + "  public String toString() {\n"
            + "    return \"Tristate{\" + \"value=\" + value + \", isNull=\" + isNull + '}';\n"
            + "  }\n"
            + "\n"
            + "  public interface OnNull<R> {\n"
            + "    /**\n"
            + "     * Registers a {@link java.util.function.Supplier} which is called in case the\n"
            + "     * property was null.\n"
            + "     */\n"
            + "    OnAbsent<R> onNull(Supplier<R> onNull);\n"
            + "  }\n"
            + "\n"
            + "  public interface OnAbsent<R> {\n"
            + "    /**\n"
            + "     * Registers a {@link java.util.function.Supplier} which is called in case the\n"
            + "     * property was absent.\n"
            + "     */\n"
            + "    R onAbsent(Supplier<R> onAbsent);\n"
            + "  }\n"
            + "}",
        writer.asString());
  }
}
