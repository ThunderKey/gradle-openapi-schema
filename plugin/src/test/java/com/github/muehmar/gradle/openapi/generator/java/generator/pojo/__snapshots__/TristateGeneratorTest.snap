tristateClass=[
java.util.Objects
java.util.Optional
java.util.function.Function
java.util.function.Supplier

package com.github.muehmar.openapi.util;

import java.util.Objects;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Supplier;

public final class Tristate<T> {
  private final Optional<T> value;
  private final boolean isNull;

  private Tristate(Optional<T> value, boolean isNull) {
    this.value = value;
    this.isNull = isNull;
  }

  public static <T> Tristate<T> ofNullableAndNullFlag(T nullableValue, boolean isNull) {
    return new Tristate<>(Optional.ofNullable(nullableValue), isNull);
  }

  public static <T> Tristate<T> ofValue(T value) {
    return new Tristate<>(Optional.of(value), false);
  }

  public static <T> Tristate<T> ofAbsent() {
    return new Tristate<>(Optional.empty(), false);
  }

  public static <T> Tristate<T> ofNull() {
    return new Tristate<>(Optional.empty(), true);
  }

  /**
   * Returns a Tristate class whose value is mapped with the given function.
   */
  public <R> Tristate<R> map(Function<T, R> f) {
    return new Tristate<>(value.map(f), isNull);
  }

  /**
   * Converts to an {@link Optional} mapping the cases if the value is null or absent
   * to an empty {@link Optional}.
   */
  public Optional<T> toOptional() {
    return onValue(Optional::of).onNull(Optional::empty).onAbsent(Optional::empty);
  }

  /**
   * Registers a {@link Function} which is applied on the value of the property if it
   * was present and non-null.
   */
  public <R> OnNull<R> onValue(Function<T, R> onValue) {
    return onNull ->
      onAbsent -> value.map(onValue).orElseGet(() -> isNull ? onNull.get() : onAbsent.get());
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Tristate<?> tristate = (Tristate<?>) o;
    return isNull == tristate.isNull && Objects.equals(value, tristate.value);
  }

  @Override
  public int hashCode() {
    return Objects.hash(value, isNull);
  }

  @Override
  public String toString() {
    return "Tristate{" + "value=" + value + ", isNull=" + isNull + '}';
  }

  public interface OnNull<R> {
    /**
     * Registers a {@link Supplier} which is called in case the property was null.
     */
    OnAbsent<R> onNull(Supplier<R> onNull);
  }

  public interface OnAbsent<R> {
    /**
     * Registers a {@link Supplier} which is called in case the property was absent.
     */
    R onAbsent(Supplier<R> onAbsent);
  }
}
]